tiled/client/catalog.py

```diff
@@ -21,6 +23,26 @@
 from .utils import handle_error
 
 
+def get_token(uri):
+    username = input("Username: ")
```

Comment:
I like making this a stand-alone function. 👍

Prompts are rare in Python APIs; they are more of a CLI thing. They do not play well with batch processing or notebook workflows. I suggest we require the username as a normal function argument (it’s not a secret) and look for the password in an env var first, `TILED_PASSWORD`, and then fall back to prompting for the password only if that is unset.

Reply:
I’ve been giving this more thought and looking around a bit, and I’d like to revise my recommendation. This function should accept a username and password directly. It’s annoying as a user to have to know a specific environment variable name or config location if I just want to get it working using

```py
f(username=os.environ["WHATEVER_I_WANT"], password= os.environ["AGAIN_WHATEVER_I_WANT"])`
```

That explicit spelling should be _option_ and the core implementation. It should come with a documented warning not to hard-code the variables...but we needn't make it _impossible_ to do so.

In later PRs, once we grow a config system, we can look to examples like https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html which provide a whole range of ways to do this which are convenient in different situations. We can add support for things like

```py
if password is None:
    password = os.getenv("TILED_PASSWORD")
```

and

```py
ClientCatalog.from_profile("CSX")  # finds a section named "CSX" in a config file containing URI, username, password, and options
```

Reply:
If we pass username, and password as function arguments. Then people will leave their password laying around in python code in plain text.

Reply:
also it seems weird to stick your plaintext password in an environment variable

Reply:
> If we pass username, and password as function arguments. Then people will leave their password laying around in python code in plain text.

I think this is a documentation issue. Higher layers, to be added _soon_ but now _now_ can provide indirect ways of injecting the password, include interactive prompting. But I really don't like the _lowest_ layer building in a prompt and assuming a TTY.

> also it seems weird to stick your plaintext password in an environment variable. Because it is not an API key, and it's a users password, this may be a password that they reuse in other places

That's a fair point, but many services including PyPI work this way. Using a token is better practice, but when we have username/password it's pretty standard to accept it via env var. Perhaps the best, most mainstream balance of convenience and security here is to support password-via-env-var but guide people toward OAuth2- or SAML-based Authenticators in the long run, where they'll be generating a code instead of handing us a password directly. Basically how Globus works from the CLI.

I pinged @dylanmcreynolds on Slack for a second opinion.

Reply:
@dylanmcreynolds is with @gwbischof that password-in-env-var is too dodgy. (As I refresh my memory..PyPI _does_ do this but [it is a known problem](https://glyph.twistedmatrix.com/2017/10/careful-with-that-pypi.html).)

I propose this, then, which is just a cosmetic refinement on what @gwbischof proposed:

```python
>>> from tiled.client.catalog import generate_token, Catalog
>>> URI = "http://..."
>>> token = generate_token(URI)
username: dallan
password: [hidden input]
>>> catalog = Catalog.from_URI(URL, token)
```

with an option to do the "text adventure" prompts in the shell:

```
$ tiled token generate http://
username:
passsword:
Token:
XXX
This token will expire in X minutes at <time>.
$ python
>>> client = Catalog.from_uri("...", token="...")
```

Reply:
(Fine with me if the CLI is a separate PR, btw. Would need a special header in the token response so that the client can know the expiration time without installing jwt and encryption libraries.)